src/sentry/api/endpoints/organization_events_facets_performance.py

```diff
@@ -72,3 +76,165 @@ def data_fn(offset, limit):
+
+    with sentry_sdk.start_span(op="discover.discover", description="facets.frequent_tags"):
+        # Get the most relevant tag keys
+        tag_data = discover.query(
+            selected_columns=initial_selected_columns,
+            query=filter_query,
+            params=params,
+            orderby=["-count"],
+            referrer="{}.{}".format(referrer, "all_transactions"),
+        )
+        counts = [r["count"] for r in tag_data["data"]]
+        aggregates = [r["aggregate"] for r in tag_data["data"]]
+
+        # Return early to avoid doing more queries with 0 count transactions or aggregates for columns that dont exist
+        if len(counts) != 1 or counts[0] == 0 or aggregates[0] is None:
+            return None
+    if not tag_data["data"][0]:
+        return None
+    return tag_data["data"][0]
+
+
+def query_facet_performance(
+    params: Mapping[str, str],
+    tag_data: Mapping[str, object],
+    aggregate_column: Optional[str] = None,
+    filter_query: Optional[str] = None,
+    orderby: Optional[str] = None,
+    referrer: Optional[str] = None,
+    limit: Optional[int] = None,
+    offset: Optional[int] = None,
+):
+    with sentry_sdk.start_span(
+        op="discover.discover", description="facets.filter_transform"
+    ) as span:
+        span.set_data("query", filter_query)
+        snuba_filter = discover.get_filter(filter_query, params)
+
+        # Resolve the public aliases into the discover dataset names.
+        snuba_filter, translated_columns = discover.resolve_discover_aliases(snuba_filter)
+
+    # Aggregate for transaction
+    transaction_aggregate = tag_data["aggregate"]
+
+    # Dynamically sample so at least 50000 transactions are selected
+    transaction_count = tag_data["count"]
+    sampling_enabled = transaction_count > 50000
+
+    # log-e growth starting at 50,000
+    target_sample = max(
+        50000 * (math.log(transaction_count) - (math.log(50000) - 1)), transaction_count
+    )
+
+    dynamic_sample_rate = 0 if transaction_count <= 0 else (target_sample / transaction_count)
+    sample_rate = min(max(dynamic_sample_rate, 0), 1) if sampling_enabled else None
+    frequency_sample_rate = sample_rate if sample_rate else 1
+
+    excluded_tags = [
+        "tags_key",
+        "NOT IN",
+        ["trace", "trace.ctx", "trace.span", "project", "browser", "celery_task_id", "url"],
+    ]
+
+    with sentry_sdk.start_span(op="discover.discover", description="facets.aggregate_tags"):
+        span.set_data("sample_rate", sample_rate)
+        span.set_data("target_sample", target_sample)
+        conditions = snuba_filter.conditions
+        aggregate_comparison = transaction_aggregate * 1.005 if transaction_aggregate else 0
+        having = [excluded_tags]
+        having.append(["aggregate", ">", aggregate_comparison])
+
+        if orderby is None:
+            orderby = []
+        else:
+            orderby = [orderby]
```

Comment:
nit: cause there's typing hinting on the orderby being passed
```suggestion
        if orderby is None:
            resolved_orderby: List[str] = []
        else:
            resolved_orderby: List[str] = [orderby]
```